spaghetto/src/Spago/Command/Ls.purs

```diff
@@ -1,3 +1,120 @@
-module Spago.Command.Ls where
+module Spago.Command.Ls (listPackages, listPackageSet, LsEnv(..), JsonFlag(..)) where
 
-import Prelude
+import Spago.Prelude
+
+import Data.Array (replicate)
+import Data.Codec.Argonaut as CA
+import Data.Codec.Argonaut.Record as CAR
+import Data.Foldable (elem, maximum)
+import Data.Map (filterKeys)
+import Data.Map as Map
+import Data.Set (unions)
+import Data.String.CodeUnits (fromCharArray, length)
+import Data.Tuple.Nested (type (/\))
+import Registry.PackageName as PackageName
+import Registry.Version as Version
+import Spago.Bin.Flags (IncludeTransitive(..))
+import Spago.Config (Dependencies(..), Package(..), PackageSet, Workspace, getPackageLocation)
+import Spago.Config as Config
+import Spago.Json (stringifyJson)
+
+data JsonFlag = JsonOutputNo | JsonOutputYes
+
+type LsEnv =
+  { dependencies :: PackageSet
+  , logOptions :: LogOptions
+  , workspace :: Workspace
+  }
+
+type JsonPackageOutput =
+  { packageName :: String
+  -- TODO: Was Json Value, has tag
+  -- See https://github.com/purescript/spago/blob/ec22308d181340382494e5bd6ef9102a3c87c0bb/src/Spago/Types.hs#L38-L47
+  , repo :: String
+  , version :: String
+  }
+
+jsonPackageOutputCodec :: JsonCodec JsonPackageOutput
+jsonPackageOutputCodec = CAR.object "JsonPackageOutput"
+  { packageName: CA.string
+  , repo: CA.string
+  , version: CA.string
+  }
+
+listPackageSet :: JsonFlag -> Spago LsEnv Unit
+listPackageSet jsonFlag = do
+  logDebug "Running `listPackageSet`"
+  { workspace } <- ask
+  output $ formatPackageNames jsonFlag (Map.toUnfoldable workspace.packageSet)
+
+listPackages :: IncludeTransitive -> JsonFlag -> Spago LsEnv Unit
+listPackages packagesFilter jsonFlag = do
+  logDebug "Running `listPackages`"
+  { dependencies, workspace } <- ask
+  let
+    packagesToList = Map.toUnfoldable case packagesFilter of
+      IncludeTransitive -> dependencies
+      NoIncludeTransitive ->
+        let
+          workspacePackages = Config.getWorkspacePackages workspace.packageSet
+          toDependencyNames (Dependencies deps) = Map.keys deps
+          directDependencies = unions $ ((toDependencyNames <<< _.dependencies <<< _.package) <$> workspacePackages)
+        in
+          filterKeys (_ `elem` directDependencies) dependencies
+  case packagesToList of
+    [] -> logWarn "There are no dependencies listed in your spago.dhall"
+    _ -> output $ formatPackageNames jsonFlag packagesToList
+
+formatPackageNames :: forall a. JsonFlag -> Array (PackageName /\ Package) -> OutputFormat a
+formatPackageNames = case _ of
+  JsonOutputYes -> OutputLines <<< formatPackageNamesJson
```

Comment:
We have `OutputJson` for these cases (the old format used json-lines, but we should just use standard JSON everywhere)

Reply:
Okay, if it doesn't need to be the same I can use `OutputJson` and `OutputTable`